File: 2020/rust/src/day6.rs
use std::collections::{HashMap, HashSet};
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day6(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 6: Custom Customs");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(input.split("\n\n").fold(0, |sum, group| {
        sum + group
            .lines()
            .fold(HashSet::new(), |set, line| {
                line.chars().fold(set, |mut qs, c| {
                    qs.insert(c);
                    qs
                })
            })
            .len() as i32
    }))
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(input.split("\n\n").fold(0, |sum, group| {
        let group_size = group.lines().count();
        sum + group
            .lines()
            .fold(HashMap::new(), |map, line| {
                line.chars().fold(map, |mut qs, c| {
                    let counter = qs.entry(c).or_insert(0);
                    *counter += 1;
                    qs
                })
            })
            .iter()
            .filter(|&(_, v)| *v == group_size)
            .count() as i32
    }))
}

File: 2020/go/day6.go
package advent2020

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func Day6(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 6: Custom Customs")
	part1, err := day6Part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := day6Part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func day6Part1(input string) (int, error) {
	sum := 0
	for _, group := range strings.Split(input, "\n\n") {
		qs := make(map[rune]struct{})
		for _, line := range strings.Split(group, "\n") {
			for _, c := range line {
				qs[c] = struct{}{}
			}
		}
		for range qs {
			sum++
		}
	}
	return sum, nil
}

func day6Part2(input string) (int, error) {
	sum := 0
	for _, group := range strings.Split(input, "\n\n") {
		qs := make(map[rune]int)
		groups := strings.Split(group, "\n")
		for _, line := range strings.Split(group, "\n") {
			for _, c := range line {
				if _, ok := qs[c]; !ok {
					qs[c] = 0
				}
				qs[c]++
			}
		}
		for _, v := range qs {
			if v == len(groups) {
				sum++
			}
		}
	}
	return sum, nil
}
